src/lsm/grid.zig

```diff
@@ -214,20 +227,67 @@ pub fn GridType(comptime Storage: type) type {
             for (grid.read_iop_tracer_slots) |slot| assert(slot == null);
             for (grid.write_iop_tracer_slots) |slot| assert(slot == null);
 
-            grid.cache.deinit(allocator);
-
             for (grid.cache_blocks) |block| allocator.free(block);
             allocator.free(grid.cache_blocks);
 
+            grid.cache_valid.deinit(allocator);
+            grid.cache.deinit(allocator);
+
             grid.* = undefined;
         }
 
+        pub fn cancel(grid: *Grid, callback: fn (*Grid) void) void {
+            assert(grid.canceling == null);
+
+            grid.canceling = .{ .callback = callback };
+
+            grid.read_pending_queue.reset();
+            grid.superblock.storage.reset_next_tick_lsm();
+            grid.superblock.storage.on_next_tick(
+                .vsr,
+                cancel_tick_callback,
+                &grid.canceling_next_tick,
+            );
+        }
+
+        fn cancel_tick_callback(next_tick: *NextTick) void {
+            const grid = @fieldParentPtr(Grid, "canceling_next_tick", next_tick);
+            if (grid.canceling == null) return;
```

Comment:
Where is `canceling` set back to null out of band? I looked through Grid but it didn't jump out?

Reply:
`cancel_join_callback`, i.e. when cancel completes